src/twister2/fixtures/builder.py

```diff
@@ -18,15 +18,22 @@ def builder(request: pytest.FixtureRequest) -> BuilderAbstract:
     """Build hex files for test suite."""
     twister_config: TwisterConfig = request.config.twister_config
     function = request.function
-    builder_klass = BuilderFactory.get_builder('west')
-    builder = builder_klass(zephyr_base=twister_config.zephyr_base, source_dir=function.spec.path)
-    function.spec.build_dir = (Path(twister_config.output_dir) / function.spec.platform
-                               / function.spec.rel_to_base_path / function.spec.original_name)
-    function.spec.build_dir = function.spec.build_dir.resolve()
-    builder.build(
-        platform=function.spec.platform,
-        build_dir=function.spec.build_dir,
-        scenario=function.spec.original_name,
-        cmake_args=function.spec.extra_args,
-    )
+    function.spec.output_dir = Path(twister_config.output_dir).resolve()
+
+    builder_class = BuilderFactory.get_builder('west')
+    builder = builder_class(zephyr_base=twister_config.zephyr_base, source_dir=function.spec.source_dir)
+
+    if not hasattr(request.session, 'build_dir_list'):
```

Comment:
This solution will work as long as tests will be run one by one. When tests will be run in parallel (with `xdist`), then information stored on session level are not shared between workers/processes, so we need more sophisticated way for keeping those information (for example in external file). Maybe it is not something to do in this PR, so please add here at least comment like: `FIXME`, or `TODO` with information that "this mechanism do not work during running tests in parallel, so it need to be refactored to be more robust."

Reply:
I added `SmartBuilderPlugin` to handle information about all built sources.